client/src/Services/useSocketCommunication.ts

```diff
@@ -1,91 +1,169 @@
   const [consoleChunks, setConsoleChunks] = useState<string[]>([]);
   const { updateCurrFocusedTab } = useGlobalStore();
+  const { setToastMessage: setMessage } = useToastStateStore();
 
-  useEffect(() => {
-    const handlers: EventHandlers = {
-      mainDebug: (data: 'Finished mainDebug event on server') => {
-        console.error(data);
-        setActiveSession(true);
-      },
-      sendFunctionDeclaration: (data: FunctionStructure) => {
-        // TODO: Implement
-        console.error('Received functional structure', data);
-      },
-      sendTypeDeclaration: (type: BackendTypeDeclaration) => {
-        updateTypeDeclaration(type);
-      },
-      sendBackendStateToUser: (state: BackendState) => {
-        updateNextFrame(state);
-      },
-      sendStdoutToUser: (output: string) => {
-        setConsoleChunks((prev) => [...prev, output]);
-      },
-      programWaitingForInput: (_data: any) => {
-        // TODO: Implement, to Hindie, might give a popup message to user?
-      },
-      compileError: (errors: string[]) => {
-        setConsoleChunks((prev) => [...prev, ...errors]);
-        updateCurrFocusedTab('2');
-      },
-      send_stdin: (data: string) => console.log('Stdin Sent:', data),
-    };
-
-    socketClient.setupEventHandlers(handlers);
+  if (!initialized) {
+    useMemo(() => {
+      const eventHandler: ServerToClientEvent = {
+        mainDebug: (_data: 'Finished mainDebug event on server') => {
+          setMessage({
+            content: 'Debug session started.',
+            colorTheme: 'info',
+            durationMs: DEFAULT_MESSAGE_DURATION,
+          });
+          setActive(true);
+        },
+        sendFunctionDeclaration: (_data: FunctionStructure) => {},
+        sendTypeDeclaration: (type: BackendTypeDeclaration) => {
+          updateTypeDeclaration(type);
+        },
+        sendBackendStateToUser: (state: BackendState | ProgramEnd) => {
+          if (isProgramEnd(state)) {
+            setMessage({
+              content: 'Debug session ended.',
+              colorTheme: 'info',
+              durationMs: DEFAULT_MESSAGE_DURATION,
+            });
+            setActive(false);
+            return;
+          }
+          updateNextFrame(state);
+        },
+        sendStdoutToUser: (output: string) => {
+          setConsoleChunks((prev) => [...prev, output]);
+        },
+        programWaitingForInput: (_data: any) => {
+          // Implement as needed
+        },
+        compileError: (errors: string[]) => {
+          setConsoleChunks((prev) => [...prev, ...errors]);
+          updateCurrFocusedTab('2');
+        },
+        send_stdin: (_data: string) => {},
+      };
 
-    return () => {
-      socketClient.clearEventHandlers(handlers);
-    };
-  }, [socketClient]);
+      socketClient.setupEventHandlers(eventHandler);
+    }, []);
+    initialized = true;
+  }
 
   const resetDebugSession = useCallback(() => {
```

Comment:
just wondering why u use so many useCallback/useMemo. Are they all resource-intensive tasks?

Reply:
UseCallback and useMemo help prevent these re-renders, it's always good to use no? 

Also, I need the functions below to be updated whenever the dependencies changes.